### PR TITLE
[ecmascript] (#361) Add Object Support in the VROES.Shims.arrayFrom()

### LINKED ISSUE #361: Shim for Array.from() doesn't handle Object type
### Description

`Array.from` can handle array-like objects, but the shim currently doesn't cover Object types and just returns the given object. This behavior also violates its type definition `any[]`, since it returns an object.

I believe adding the following case to the shim restores the basic behavior to match `Array.from`

```ts
case "Object":
  arrayLikeClone = []
  if (Object.hasOwn(arrayLike, 'length')) {
      arrayLikeClone = Array.apply(null, Array(arrayLike.length))
  }
  break;
```

Examples of the actual function can be seen [here on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from#using_arrow_functions_and_array.from).

### Steps to Reproduce

```ts
// Actual function:
Array.from({ length: 5 })
// [undefined, undefined, undefined, undefined, undefined]
Array.from({ length: 5 }, (_, i) => i);
// [0, 1, 2, 3, 4]
Array.from({ foo: 'bar' })
// []

// Shim behavior
arrayFrom({ length: 5 });
// { length: 5 }
arrayFrom({ length: 5 }, (_, i) => i);
// { length: 5 }
arrayFrom({ foo: 'bar' })
// { foo: 'bar' }
```

**Preconditions:** none

**Expected behavior:** The shim to handle Object types more like the `Array.from` function

**Actual behavior:** Just returns the same object back

**Reproduces how often:** 100%

**Component/s:** `ecmascript/Shims`

**Affects Build/s:** <= 2.40.0

#### Client

-   Build Tools for VMware Aria Version: 2.40.0
-   Visual Studio Code Version: n/a
-   OS Version: n/a

#### Server

-   vRealize Automation Version: n/a
-   vRealize Orchestrator Version: n/a
-   vRealize Operations Version: n/a
-   vRealize Log Insight Version: n/a

### PR DESCRIPTION
**Pull request by akantchev**
### Description

Add support of objects in the `VROES.Shims.arrayFrom()` method so its behavior is similar to the standard `Array.from()` method.

### Checklist

- [x] I have added necessary documentation, relevant usage information (if applicable)
- [x] I have updated the PR title with affected component, related issue number and a short summary of the changes introduced
- [x] I have added labels for implementation kind (kind/*) and version type (version/*)
- [x] I have tested against live environment, if applicable
- [x] I have my changes rebased and squashed to the minimal number of relevant commits. **Notice: don't squash all commits**
- [x] I have added a descriptive commit message with a short title, including a `Fixed #XXX -` or `Closed #XXX -` prefix to auto-close the issue

### Testing

1. Create a workflow with one script able task that has the following code:
```js
var __global = System.getContext() || (function () { return this; }).call(null);
var VROES = __global.__VROES || (__global.__VROES = System.getModule("com.vmware.pscoe.library.ecmascript").VROES()), exports = {};

var array1 = array1 = VROES.Shims.arrayFrom({ length: 5 });
System.log("Array 1  data: " + JSON.stringify(array1, null, 4));

var arr = [{ key1: 1 }, { key2: 2 }, 3, "ddd"];
var array2 = VROES.Shims.arrayFrom(arr);
System.log("Array 2  data: " + JSON.stringify(array2, null, 4));
```
2. Run the workflow, the output should be array based on the input object.

#### Test Screenshot

361_array_test_results

### Release Notes

Add support of objects in the `VROES.Shims.arrayFrom()` method so its behavior is similar to the standard `Array,from()` method.

### Related issues and PRs

Resolves #361

### CODE REVIEW

File: docs/versions/latest/Release.md [hunk 1]
```diff
@@ -497,18 +533,18 @@ export class HandleNetworkConfigurationBackup {
...
-[//]: # (Optional But higlhy recommended Specify *NONE* if missing)
-[//]: # (#### Relevant Documentation:)
+[//]: # "### *Improvement Name* "
+[//]: # "Talk ONLY regarding the improvement"
+[//]: # "Optional But higlhy recommended"
+[//]: # "#### Previous Behavior"
+[//]: # "Explain how it used to behave, regarding to the change"
+[//]: # "Optional But higlhy recommended"
+[//]: # "#### New Behavior"
+[//]: # "Explain how it behaves now, regarding to the change"
```
Comment (VenelinBakalov) [inline docs/versions/latest/Release.md:543]: For this kind of changes ( ) to " " and * * to _ _ could you double check with @ktsonchev , he started updating documentation for similar things and I want to double check that everything is alligned

File: packages/ecmascript/src/Shims.ts [hunk 2]
```diff
@@ -124,30 +124,32 @@ export default class Shims {
...
-		if (mapfn) {
-			array = arrayLikeClone.map(mapfn);
-		} else {
-			array = arrayLikeClone;
+			case "Object":
+				arrayLikeClone = [];
+				// check whether the object is an array (using the length property)
+				if (Object.keys(arrayLike).find(item => item.indexOf('length') >= 0)) {
+					arrayLikeClone = Array.apply(null, Array(arrayLike));// nosonar
+				}
```
Comment (Indy-rbo) [inline packages/ecmascript/src/Shims.ts:132]: I don't mind a rewritten version, but this does not do what I suggested!
```suggestion
				if (Object.hasOwn(arrayLike, 'length')) {
					arrayLikeClone = Array.apply(null, Array(arrayLike.length))
				}
```

Running both functions gives:
```
arrayFromPR({ length: 5 }, (v, i) => i); // Output: [0]
arrayFromPR({ length: 5 }) // Output: [{"length": 5}]
arrayFromMySuggestion({ length: 5 }, (v, i) => i); // Output: [0,1,2,3,4]
arrayFromMySuggestion({ length: 5 }) // Output: [undefined,undefined,undefined,undefined,undefined]
```

It does make it a bit more "compliant" to its type signature in that it now gives an array of the object. But it does not behave like the actual implementation.

**Review by Indy-rbo — CHANGES_REQUESTED** [review]
This change does not resolve my issue

File: packages/ecmascript/src/Shims.ts (hunk 2 above)
Comment (Indy-rbo) [reply packages/ecmascript/src/Shims.ts:132]: Re-reading the snippet, it seems you just missed the `.length`! :)
`arrayLikeClone = Array.apply(null, Array(arrayLike.length))`
`arrayLikeClone = Array.apply(null, Array(arrayLike));// nosonar`

File: docs/versions/latest/Release.md [hunk 3]
```diff
@@ -1,34 +1,37 @@
-[//]: # (VERSION_PLACEHOLDER DO NOT DELETE)
-[//]: # (Used when working on a new release. Placed together with the Version.md)
-[//]: # (Nothing here is optional. If a step must not be performed, it must be said so)
-[//]: # (Do not fill the version, it will be done automatically)
-[//]: # (Quick Intro to what is the focus of this release)
+[//]: # "VERSION_PLACEHOLDER DO NOT DELETE"
```
Comment (Michaelpalacce) [inline docs/versions/latest/Release.md:1]: Lets revert these changes

File: docs/versions/latest/Release.md [hunk 4]
```diff
@@ -1,34 +1,37 @@
...
+### Support of Objects in the VROES.Shims.arrayFrom() Method
+
+Add support for objects in the `VROES.Shims.arrayFrom()` method so its behavior is similar to the standard `Array.from()` method.
 
 ### Support python 3.10 runtime
 
 Add support for python 3.10 runtime in Orchestrator. This is now the default, since python 3.7 is deprecated.
 
-### *New `@ActionItem` decorator for Workflows*
+### _New `@ActionItem` decorator for Workflows_
```
Comment (Michaelpalacce) [inline docs/versions/latest/Release.md:34]: This too

File: docs/versions/latest/Release.md [hunk 5]
```diff
@@ -214,40 +220,51 @@ Special output is needed for the AsyncWorkflowItem.
 
 No special inputs are needed for the AsyncWorkflowItem.
 
-
 #### Example
 
 ```typescript
-import { Workflow, Out, In, Item, RootItem, AsyncWorkflowItem } from "vrotsc-annotations";
+import {
```
Comment (Michaelpalacce) [inline docs/versions/latest/Release.md:226]: I don't know what kind of auto formatter you are using, but these kind of changes everywhere are not needed. 

1. trailing commas don't make the code better
2. the line width you have set up should really be introduced.... we no longer have small screens for this wrapping to make sense.
3. Let's revert everything not related to your PR please

**Review by Michaelpalacce — CHANGES_REQUESTED** [review]

File: docs/versions/latest/Release.md (hunk 3 above)
Comment (akantchev) [reply docs/versions/latest/Release.md:1]: Discussed with @VenelinBakalov , no need for reverting currently.

File: docs/versions/latest/Release.md (hunk 5 above)
Comment (akantchev) [reply docs/versions/latest/Release.md:226]: Discussed with @VenelinBakalov , no need for reverting currently.

File: docs/versions/latest/Release.md (hunk 4 above)
Comment (akantchev) [reply docs/versions/latest/Release.md:34]: Discussed with @VenelinBakalov , no need for reverting currently.

File: packages/ecmascript/src/Shims.ts (hunk 2 above)
Comment (akantchev) [reply packages/ecmascript/src/Shims.ts:132]: Thank you for your suggestion, however this does not compile with the older version of the `tslib`, hence the code is written in this way.

File: docs/versions/latest/Release.md (hunk 1 above)
Comment (akantchev) [reply docs/versions/latest/Release.md:543]: Discussed offline.

File: packages/ecmascript/src/Shims.ts (hunk 2 above)
Comment (Indy-rbo) [reply packages/ecmascript/src/Shims.ts:132]: Ah alright. I understand that the `Object.hasOwn` might not work, but without calling `Array.apply(null, Array(arrayLike.length))` this does not behave correctly. You called `Array(arrayLike)` not `Array(arrayLike.length)`

File: docs/versions/latest/Release.md (hunk 3 above)
Comment (VenelinBakalov) [reply docs/versions/latest/Release.md:1]: There is no urgency to merge, let's wait for @ktsonchev to come back and we can cover this with him to make sure everyone is alligned.

**Review by VenelinBakalov — CHANGES_REQUESTED** [review]
Pending review on the documentation changes by @ktsonchev .
In the meantime, @Michaelpalacce and @akantchev please double check if there are any other changes required purely from code perspective and we will align next week.

File: docs/versions/latest/Release.md (hunk 5 above)
Comment (VenelinBakalov) [reply docs/versions/latest/Release.md:226]: @Michaelpalacce - we discussed that we can merge the current changes suggested by the autoformater (the formater should be handling some sonar/linting issues so might be a good approach to introduce) and next week we will check with Kalin and if something is not ok, we can align the settings and revert.
However, after that discussion a new issue popped up that might require a release this week. Because of that it would be best to either postpone this PR for next week or revert the changes and merge without them until we have covered the discussion with Kalin.

File: packages/ecmascript/src/Shims.ts (hunk 2 above)
Comment (akantchev) [reply packages/ecmascript/src/Shims.ts:132]: Hi actually if you take a look at the screenshot it is behaving correctly with various types of objects. The `length` parameter is actually not required.

File: docs/versions/latest/Release.md (hunk 5 above)
Comment (akantchev) [reply docs/versions/latest/Release.md]: The changes were reverted, so we can merge this one as well.

File: packages/ecmascript/src/Shims.ts (hunk 2 above)
Comment (VenelinBakalov) [reply packages/ecmascript/src/Shims.ts:132]: Hi @Indy-rbo we discussed the changes internally, there is some difference in the behaviour, we will check it.

Comment (VenelinBakalov) [reply packages/ecmascript/src/Shims.ts:132]: Leaving an example to make things easier to understand:
```javascript
console.log(Array.from({ 0: 132, 2: "asd", length: 5 }));
// Logs: "[132, undefined, "asd", undefined, undefined]"
```

Comment (Indy-rbo) [reply packages/ecmascript/src/Shims.ts:132]: As I was writing this out I saw the updates. It also includes a suggested change, so to hopefully assist I'll still post it:

---

Hi, I'd like to explain that this currently does not do what is expected.

Example 1 shows that `VROES.Shims.arrayFrom({ length: 5 });` gives `[{ length: 5 }]`. But  using the actual function it gives `[undefined, undefined, undefined, undefined, undefined]`

image

Example 2 is not valid, since you are actually passing in an Array which contains an object (and a number and a string). So that example does not hit the `Object` case.


The comment indicates there might be a misunderstanding of what my snippet did. 

`// check whether the object is an array (using the length property)`
We are not checking the length to check if the object is somehow also an array. We check the length since it's a special property for array-like objects. As the MDN docs say "array-like objects (objects with a length property and indexed elements)"

Reading the MDN documentation might help getting a clearer understanding of what it should do https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from

---

After reading it further myself, it turns out my suggestion was naive as the following case won't be handled:

```ts
Array.from({
    1: 'foo',
    2: 'bar',
    'non-indexed': 'baz',
    length: 5
})
```
Which should return:
`[undefined, 'foo', 'bar', undefined, undefined]`
image


So my adjusted snippet would be:

```ts
case "Object":
	arrayLikeClone = [];
	// Check if the object is an array-like object
	if (Object.keys(arrayLike).find(item => item.indexOf('length') >= 0)) {
		arrayLikeClone = Array.apply(null, Array(arrayLike.length));// nosonar
		Object.keys(arrayLike).forEach(element => {
			const indexKey = parseInt(element);
			if (!isNaN(indexKey) && indexKey < arrayLike.length) { // Check if object key is like an indexed element and within range
				arrayLikeClone[element] = arrayLike[element];
			}
		});
	}
	break;
```

Which in simple testing already came closer to the expected result:

image

Also if the index key is out of bounds it behaves as expected:
image

Comment (VenelinBakalov) [reply packages/ecmascript/src/Shims.ts:132]: Probably the easiest workaround here would be to create a new Array with the given length and after that add the objects for the given index:
```typescript
		const arrayLength = (arrayLike as any).length;
		arrayLikeClone = new Array(arrayLength);
		for (let i = 0; i < arrayLength; i++) {
			array[i] = arrayLikeClone[i];
		}
```
Something of this kind should do the magic.

@akantchev @Indy-rbo what do you think?

Comment (VenelinBakalov) [reply packages/ecmascript/src/Shims.ts:132]: @Indy-rbo I just saw your message above, yes, this is how I understand it as well. I think your code sample is better as it has more safety checks and should do the job, thank you!
@akantchev could you check the snippet from the reply just above mine? (https://github.com/vmware/build-tools-for-vmware-aria/pull/362#discussion_r1698241172)

Comment (akantchev) [reply packages/ecmascript/src/Shims.ts]: Hi @Indy-rbo , @VenelinBakalov.
After short internal discussion we figured out the exact issue and what code update is needed, code is updated to have exactly the result you mentioned.

File: packages/ecmascript/src/Shims.ts [hunk 6]
```diff
@@ -16,10 +16,10 @@ function getPadString(sourceLength: number, targetLength: number, padString?: st
 	if (sourceLength > targetLength) {
 		return "";
 	}
-	if (padString != null) {
+	if (padString !== null) {
 		padString = "" + padString;
 	}
-	if (padString == null || !padString.length) {
+	if (padString !== null || !padString.length) {
```
Comment (VenelinBakalov) [inline packages/ecmascript/src/Shims.ts:22]: I just saw this one, I think there should not be a negation, otherwise we are changing behaviour:
```suggestion
	if (padString === null || !padString.length) {
```

File: packages/ecmascript/src/Shims.ts (hunk 2 above)
Comment (VenelinBakalov) [reply packages/ecmascript/src/Shims.ts]: @Indy-rbo I think we Aleks' latest changes we should be good. Whenever you have a chance, please take a look as well
@akantchev could you just check one minor comment on the other changed file

File: packages/ecmascript/src/Shims.ts (hunk 6 above)
Comment (akantchev) [reply packages/ecmascript/src/Shims.ts]: Hi @VenelinBakalov, thanks for the catch, code updated

**Review by VenelinBakalov — APPROVED** [review]

**Review by Indy-rbo — APPROVED** [review]

**Comment by Indy-rbo** [comment]
Thanks everyone! This should now correctly resolve #361

File: packages/ecmascript/src/Shims.ts [hunk 7]
```diff
@@ -124,30 +124,41 @@ export default class Shims {
...
+						const indexKey = parseInt(element);
+						// Check if object key is like an indexed element and within range
+						if (!isNaN(indexKey) && indexKey < length) {
+							arrayLikeClone[element] = arrayLike[element];
+						}
+					});
+				}
+				break;
+			default:
+				return arrayLikeClone;
```
Comment (Indy-rbo) [inline packages/ecmascript/src/Shims.ts:144]: Oh something that came to mind later. Maybe default to returning an empty array? 
If the shim can't handle it, it should still honor its return type (which is any[])
```suggestion
				return [];
```

Simple example:

```ts
arrayFrom(10) // Output: 10
Array.from(10) // Output: []
```

**Review by Indy-rbo — CHANGES_REQUESTED** [review]
Sorry something came to mind later regarding honoring the functions return type

**Comment by VenelinBakalov** [comment]
> Thanks everyone! This should now correctly resolve #361

thank you for your suggestions and support as well!

File: packages/ecmascript/src/Shims.ts (hunk 7 above)
Comment (VenelinBakalov) [reply packages/ecmascript/src/Shims.ts:144]: I will take care of it in separate PR

Comment (VenelinBakalov) [reply packages/ecmascript/src/Shims.ts:144]: @Indy-rbo @akantchev FYI:
https://github.com/vmware/build-tools-for-vmware-aria/pull/372
